src/Common/CARESPTRResolver.cpp

```diff
@@ -0,0 +1,77 @@
+#include "CARESPTRResolver.h"
+#include <Common/Exception.h>
+#include <arpa/inet.h>
+#include "netdb.h"
+#include "ares.h"
+
+namespace DB {
+
+    namespace ErrorCodes
+    {
+        extern const int DNS_ERROR;
+    }
+
+    static void callback(void * arg, int status, int, struct hostent * host) {
+        auto * ptr_records = reinterpret_cast<std::vector<std::string>*>(arg);
+        if (status == ARES_SUCCESS) {
+            int i = 0;
+            while (auto * ptr_record = host->h_aliases[i]) {
+                ptr_records->emplace_back(ptr_record);
+                i++;
+            }
+        }
+    }
+
+    CARESPTRResolver::CARESPTRResolver() : channel(std::make_unique<ares_channel>()) {
+        init();
+    }
+
+    CARESPTRResolver::~CARESPTRResolver() {
+        deinit();
+    }
+
+    std::vector<std::string> CARESPTRResolver::resolve(const std::string & ip) {
+        std::vector<std::string> ptr_records;
+
+        resolve(ip, ptr_records);
+        wait();
+
+        return ptr_records;
+    }
+
+    void CARESPTRResolver::init() {
+        if (ares_init(channel.get()) != ARES_SUCCESS) {
+            throw DB::Exception("Failed to initialize c-ares", DB::ErrorCodes::DNS_ERROR);
+        }
+    }
+
+    void CARESPTRResolver::deinit() {
+        ares_destroy(*channel);
+        ares_library_cleanup();
+    }
+
+    void CARESPTRResolver::resolve(const std::string & ip, std::vector<std::string> & response) {
+        in_addr addr;
+        inet_aton(ip.c_str(), &addr);
+
+        ares_gethostbyaddr(*channel, reinterpret_cast<const char*>(&addr), sizeof(addr), AF_INET, callback, &response);
```

Comment:
The second argument should be `const void *`, why is it reinterpreted as `const char*`? 
Also do we need to pass `AF_INET6` for IPv6?

Reply:
I need to work on the IPv6 part of it

Reply:
Just tested an impl of ipv6 in ClickHouse, it works. Will publish it soon